chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,167 @@
+package mir
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	lapi "github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/v1api"
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+	"github.com/filecoin-project/lotus/chain/types"
+	ltypes "github.com/filecoin-project/lotus/chain/types"
+	mirproto "github.com/filecoin-project/mir/pkg/pb/requestpb"
+	"github.com/libp2p/go-libp2p-core/host"
+	"go.uber.org/zap/buffer"
+)
+
+const (
+	ReconfigurationInterval = 2000 * time.Millisecond
+)
+
+// Mine implements "block mining" using the Mir framework.
+//
+// Mine implements the following main algorithm:
+// 1. Retrieve messages and cross-messages from the mempool.
+//    Note, that messages can be added into mempool via the libp2p and CLI mechanism.
+// 2. Send messages and cross messages to the Mir node through the request pool implementing FIFO.
+// 3. Receive ordered messages from the Mir node and parse them.
+// 4. Create the next Filecoin block.
+// 5. Broadcast this block to the rest of the network. Validators will not accept broadcasted,
+//    they already have it.
+//
+func Mine(ctx context.Context, addr address.Address, h host.Host, api v1api.FullNode, membershipCfg string) error {
+	log.With("addr", addr).Infof("Mir miner started")
+	defer log.With("addr", addr).Infof("Mir miner completed")
+
+	m, err := NewManager(ctx, addr, h, api, membershipCfg)
+	if err != nil {
+		return fmt.Errorf("unable to create a manager: %w", err)
+	}
+
+	log.Infof("Miner info:\n\twallet - %s\n\tsubnet - %s\n\tMir ID - %s\n\tPeer ID - %s\n\tvalidators - %v",
+		m.Addr, m.NetName, m.MirID, h.ID(), m.InitialValidatorSet.GetValidators())
+
+	mirErrors := m.Start(ctx)
+
+	reconfigure := time.NewTicker(ReconfigurationInterval)
+	defer reconfigure.Stop()
+
+	lastValidatorSet := m.InitialValidatorSet
+
+	var configRequests []*mirproto.Request
+
+	for {
+		// Here we use `ctx.Err()` in the beginning of the `for` loop instead of using it in the `select` statement,
+		// because if `ctx` has been closed then `api.ChainHead(ctx)` returns an error,
+		// and we will be in the infinite loop due to `continue`.
+		if ctx.Err() != nil {
+			log.Debug("Mir miner: context closed")
+			return nil
+		}
+		base, err := api.ChainHead(ctx)
+		if err != nil {
+			log.Errorw("failed to get chain head", "error", err)
+			continue
+		}
+
+		nextEpoch := base.Height() + 1
```

Comment:
Can we call this `nextHeight`?
Just to disambiguate wrt. Mir epochs...

Reply:
We can :) and that's actually a great point.